### ISSUE TITLE
Is there a way to get position at a specific time?

### ISSUE THREAD

**Issue by toxicaliengf** [Karth42/SimpleKeplerOrbits#3]
Thank you for making such a great tool!

I'm trying to use this to create "spheres of influence" to fake n-body gravity as discussed in [this talk by the kerbal developers](https://youtu.be/mXTxQko-JH0?t=1091). Basically I want to check an orbit after it's created to see if it enters another body's sphere of influence by looking at different points along the orbit, & comparing the distance between where the orbiter will be & where each sphere of influence will be. However I hit a snag when trying to plug a time value into the OrbitMover.

My first approach was to try plugging in different deltaTime values into OrbitData.UpdateOrbitDataByTime in order to step forward through time, then give it a negative value at the end to "reset". However I hit some unpredictable results & am not sure what's causing it- it seems correct at first but the accuracy degrades over time (or possibly depending on which part of the orbit is being sampled). I think maybe the issue is with adding negative deltaTime values?

Is there a way to sample position of an OrbitMover at (Time.time = 100 etc) instead of incrementing the system each frame? And/ or is there a way to sample the orbit mover's position by percentage?

Thanks🌼

**Comment by Karth42** [comment]
Hello. Excellent question.
There are many ways to do it, and first of all lets say that you could just clone Orbit data instance for sampling (you can use generic memeberwiseclone for this), preserving original orbit unchanged. 
However, it is not best way by far.


Regarding moving along orbit by percentage: There is no "time" in kepler orbit simulation, there is only current state of anomalies and eccentricity. 
Visual motion along orbit, which you can see when KeplerMover is updating, is just a trick, accomplished by constant progression of mean anomaly.
Mean anomaly is converted to true anomaly (and eccentric anomaly), which produces correct angle of body position. (BTW, anomalies are just angles).
The process of convertion of mean anomaly to true anomaly is main reason why visual motion is look like have proper acceleration: 
difference between true and mean anomaly is maximal when object is closer to attractor(focal point), and minimal when object is farthest from attractor, which is defined by Kepler's equation.
This brief explanation may be helpful for understanding of next proposed solutions.

As you can see, for moving along orbit by percentage, not speed, you can just change true anomaly, instead of mean anomaly. 
Value range is from 0 to 2PI, which you can divide by any amount of peaces and add to current anomaly value (as starting point).
There is a special method for setting new value:
```CS
void KeplerOrbitData.SetTrueAnomaly(double t)
```
which also automatically refreshes whole orbit state.
In other words, if you want to progress simulation to specific time in your world, you need to know 
the value of true anomaly at that moment (or any other anomaly, because if you know one anomaly, you can easily calculate other two).

And one last thing, you can bypass calculation of full orbit state when sampling orbit points by using 
```CS
Vector3d KeplerOrbitData.GetFocalPositionAtTrueAnomaly(double t)
```
this way it will be more optimal, because velocity vector is not calculated, when it is not used. 
This method is also used in orbit points calcualtion algorythm (GetOrbitPoints methods), which is very close to what you want to get.

(also should mention that dynamic sphere of influence has been implemented in my another plugin, SpaceGravity2D, which is available in asset store)

**Comment by toxicaliengf** [comment]
Thanks for this thorough response! This is really helpful. I'll tinker around with GetFocalPositionAtTrueAnomaly & also take a look at SpaceGravity2D 🌼

**Comment by Karth42** [comment]
Also try to change eccentric anomaly, if true anomaly is giving you incorrect result (I might have mixed up a little these anomalies, because as I just recall, true anomaly constant delta sections will not always have proportional distances between points)
